src/python/pants/backend/python/rules/python_test_debugger.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from pathlib import Path
+
+from pants.backend.python.rules.inject_init import InjectedInitDigest
+from pants.backend.python.rules.pex import (
+  CreatePex,
+  Pex,
+  PexInterpreterConstraints,
+  PexRequirements,
+)
+from pants.backend.python.subsystems.pytest import PyTest
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.base.build_root import BuildRoot
+from pants.build_graph.address import Address
+from pants.engine.fs import Digest, DirectoriesToMerge, DirectoryToMaterialize, Workspace
+from pants.engine.interactive_runner import InteractiveProcessRequest, InteractiveRunner
+from pants.engine.legacy.graph import BuildFileAddresses, HydratedTarget, TransitiveHydratedTargets
+from pants.engine.legacy.structs import PythonTestsAdaptor
+from pants.engine.rules import rule
+from pants.engine.selectors import Get, MultiGet
+from pants.rules.core.core_test_model import TestDebugResult
+from pants.rules.core.strip_source_root import SourceRootStrippedSources
+from pants.util.contextutil import temporary_dir
+
+
+@rule(name="Run pytest in an interactive process")
+async def debug_python_test(
+  test_target: PythonTestsAdaptor,
+  pytest: PyTest,
+  python_setup: PythonSetup,
+  build_root: BuildRoot,
+  runner: InteractiveRunner,
+  workspace: Workspace,
+) -> TestDebugResult:
+
+  transitive_hydrated_targets = await Get[TransitiveHydratedTargets](
+    BuildFileAddresses((test_target.address,))
+  )
+  all_targets = transitive_hydrated_targets.closure
+  all_target_adaptors = [t.adaptor for t in all_targets]
+
+  interpreter_constraints = PexInterpreterConstraints.create_from_adaptors(
+    adaptors=tuple(all_target_adaptors),
+    python_setup=python_setup
+  )
+
+  requirements = PexRequirements.create_from_adaptors(
+    adaptors=all_target_adaptors,
+    additional_requirements=pytest.get_requirement_strings()
+  )
+
+  source_root_stripped_test_target_sources = await Get[SourceRootStrippedSources](
+    Address, test_target.address.to_address()
+  )
+
+  source_root_stripped_sources = await MultiGet(
+    Get[SourceRootStrippedSources](HydratedTarget, hydrated_target)
+    for hydrated_target in all_targets
+  )
+
+  stripped_sources_digests = tuple(
+    stripped_sources.snapshot.directory_digest for stripped_sources in source_root_stripped_sources
+  )
+  sources_digest = await Get[Digest](DirectoriesToMerge(directories=stripped_sources_digests))
+
+  output_pytest_requirements_pex = 'pytest-with-requirements.pex'
+
+  resolved_requirements_pex = await Get[Pex](
+    CreatePex(
```

Comment:
There is a fair amount of duplication between this and the `python_test_runner` rule - we would be able to leverage [`create_pex_from_target_closure`](https://github.com/pantsbuild/pants/pull/8811/files) to replace much of this (the only mod it would need is the ability to pass the `additional_requirements` seen on l51).

Reply:
`create_pex_from_target_closure` is now in master, FWIW.

Reply:
The need to duplicate the test runner was one of the motivations for the design from https://docs.google.com/document/d/1Hn73YlhTPROlULTMa_3A-Fdv7hAPiHFII8rvXri5l7E/edit?disco=AAAAD2DjLRk : basically, having a "foreground=True" flag. cc @gshuflin

It would be really good to confirm that adding that flag would not be a net simplification over needing to duplicate rules in non-composable ways like this.

Reply:
@stuhood @codealchemy n.b. some of these comments were brought up in the context of https://github.com/pantsbuild/pants/pull/8855 

The rules necessary to implement --debug should be composeable with the exiting `run_python_test` rule that yields a `TestResult` up to here: https://github.com/pantsbuild/pants/blob/d5b1c3da236009ad0df237ded21bfa9ae724243d/src/python/pants/backend/python/rules/python_test_runner.py#L94

This line of code returns an `ExecuteProcessRequest`  and then runs it. The --debug version should take this same `ExecuteProcessRequest` and extract the fields from it that are relevant to running an `InteractiveProcessRequest`. 

`create_execute_request` in `hermetic_pex.py` returns an `ExecuteProcessRequest` that has the fields:
```
      argv=argv,
      input_files=input_files,
      description=description,
      env=hermetic_env
 ```
which are implemented on `InteractiveProcessRequest` (at least with the above PR which implements `input_files`), as well as a `timeout` (which is not currently implemented on `InteractiveProcessRequest`, but would be easy to do).

It was suggested that `InteractiveProcessRequest` should perhaps change to just wrap the `ExecuteProcessRequest` type, rather than duplicate some of its fields. If we go that route, than that makes this --debug rule ever-so-slightly easier to write because then the --debug flag can control whether or not the rule graph asks for the EPR directly or asks for an InteractiveProcessRequest that wraps the EPR. 

But either way, this test-debug rule can be implemented by abstraction all of the `run_python_test` except for the last few lines into a new rule, and then composing that rule with additional rules that get us the new behavior we want, i.e. the ability to get either an ExecuteProcessRequest or InteractiveProcessRequest based on --debug.

Reply:
> Before continuing with this implementation, it would be really good to confirm that adding that flag would not be a net simplification over needing to duplicate rules in non-composable ways like this.

I don't have a sense of how much work this would be to change. If it's significant, I think I'd prefer to get this in ~as-is (in terms of the overall design, even with the duplication), so we can start working with it, and refactor it to use a proposed --foreground flag in a follow-up, once we better understand the use case and its consequences.

As far as I can tell tests are the only place where this problem arises, so it may end up being the case that the duplication is actually OK, or even better. It may make sense to start off with it and see.

Reply:
> As far as I can tell tests are the only place where this problem arises, so it may end up being the case that the duplication is actually OK, or even better. It may make sense to start off with it and see.

+1. Don't generalize until we have enough use cases to have a solid understanding of what the generalization needs to support.

> There is a fair amount of duplication between this and the python_test_runner rule - we would be able to leverage create_pex_from_target_closure to replace much of this

I'm still concerned about the impact of caching on this. Currently, we produce a requirements PEX with no sources in it. That Pex has a high cache hit rate. Instead, `create_pex_from_target_closure` would include the source files in the generated Pex so the cache would no longer be useful between different targets as they will always have different source files.

--

Instead, consider adding this debugger implementation directly in `python_test_runner.py` and deduplicating there, similar to what we do for Black and Isort: https://github.com/pantsbuild/pants/blob/master/src/python/pants/backend/python/lint/black/rules.py and https://github.com/pantsbuild/pants/blob/master/src/python/pants/backend/python/lint/isort/rules.py.

Until after `merged_input_files = await Get(Digest, ...)`, everything seems duplicated from a quick scan. Introduce a new `PytestSetup` dataclass that contains the requirements Pex, merged input files, and CLI args, and then have the rules `test` and `debug` request a `PytestSetup` and feed it into an EPR and IPR, respectively. 

Happy to pair on this with you.

Reply:
I started prototyping the `interactive=True` flag, and so far it's pretty straightforward. Broken patch here: https://github.com/pantsbuild/pants/compare/master...twitter:stuhood/process-execute-foreground

If this PR really needs to land to support a usecase you folks have asap, then I won't block. But _please_ make it easy to deprecate the duplicated code later, as I expect that the foreground patch could be polished in a few more hours of work.

Reply:
@stuhood isn't this what InteractiveProcessRunner was meant to do, with the additional constraint that we side-step having to think about how EPR does caching by requiring that InteractiveProcessRunner only execute from console_rules?

Reply:
Yes. Except, again, InteractiveProcessRunner requires changes like this PR, which duplicate a lot of code. There would not be any code duplication in the case of `foreground=True`, because it could be used arbitrarily deep in the rule graph.